composeml/label_maker.py

```diff
@@ -4,141 +4,333 @@
+        offset = value
 
     elif isinstance(value, str):
-        offset = pd.Timedelta(value)
-        assert offset is not pd.NaT, 'invalid offset'
-        assert offset.total_seconds() >= 0, 'negative offset'
+        error = 'offset must be a valid string'
+        assert can_be_type(type=pd.tseries.frequencies.to_offset, string=value), error
+        offset = pd.tseries.frequencies.to_offset(value)
+        assert offset.n > 0, 'offset must be greater than zero'
 
     else:
-        raise TypeError('invalid offset type')
+        assert is_offset(value), 'invalid offset'
+        assert value.n > 0, 'offset must be greater than zero'
+        offset = value
+
+    return offset
 
 
 class LabelMaker:
     """Automatically makes labels for prediction problems."""
 
-    def __init__(self, target_entity, time_index, labeling_function, window_size):
-        """
-        Creates an instance of label maker.
+    def __init__(self, target_entity, time_index, labeling_function, window_size=None):
+        """Creates an instance of label maker.
 
         Args:
             target_entity (str) : Entity on which to make labels.
             time_index (str): Name of time column in the data frame.
             labeling_function (function) : Function that transforms a data slice to a label.
             window_size (str or int) : Duration of each data slice.
+                The default value for window size is all future data.
         """
-        assert_valid_offset(window_size)
-
         self.target_entity = target_entity
         self.time_index = time_index
         self.labeling_function = labeling_function
         self.window_size = window_size
 
-    def _preprocess(self, df):
-        if df.index.name != self.time_index:
-            df = df.set_index(self.time_index)
+        if self.window_size is not None:
+            self.window_size = to_offset(self.window_size)
 
-        if 'time' not in str(df.index.dtype):
-            df.index = df.index.astype('datetime64[ns]')
+    def get_slices(self, df, gap=None, min_data=None, drop_empty=True):
+        """Generate data slices.
 
-        return df
+        Args:
+            df (DataFrame) : Data frame to generate data slices.
+            gap (str) : Time between slices. Default value is window size.
+            min_data (int or str or Timestamp) : Threshold to cutoff data.
+            drop_empty (bool) : Whether to drop empty slices.
+
+        Returns:
+            DataFrame, dict : Returns a data slice and metadata about the data slice.
+        """
+        self.window_size = self.window_size or len(df)
+        gap = to_offset(gap or self.window_size)
+
+        df = df.loc[df.index.notnull()]
+        df.sort_index(inplace=True)
+
+        if df.empty:
+            return
+
+        threshold = min_data or df.index[0]
+        df, cutoff_time = cutoff_data(df=df, threshold=threshold)
+
+        if df.empty:
+            return
+
+        if isinstance(gap, int):
```

Comment:
not sure i understand why this is needed. if you set a minimum data with timestamp, we don't want it to skip ahead in time just be cause gap was an int, right?

Reply:
If a user sets the minimum data as a timestamp and the gap as relative (int), the minimum data can't be used as the first cutoff time as it would no longer be relative (e.g. based on the actual values in the time index). When using relative gaps, this ensures that the first cutoff time is also relative.

Reply:
i see. so basically, if gap is relative, it will make sure you start on the first event after the minimum data. that makes sense.

this detail should go in the doc string for minimum data

Reply:
yes, that is correct